src/shared/install.c

```diff
@@ -705,132 +705,91 @@ static int is_symlink_with_known_name(const UnitFileInstallInfo *i, const char *
 
-        _cleanup_closedir_ DIR *d = NULL;
         struct dirent *de;
         int r = 0;
 
-        assert(i);
-        assert(fd >= 0);
-        assert(path);
-        assert(config_path);
-        assert(same_name_link);
-
-        d = fdopendir(fd);
-        if (!d) {
-                safe_close(fd);
-                return -errno;
-        }
-
-        FOREACH_DIRENT(de, d, return -errno) {
-
-                dirent_ensure_type(d, de);
-
-                if (de->d_type == DT_DIR) {
-                        _cleanup_free_ char *p = NULL;
-                        int nfd, q;
+        FOREACH_DIRENT(de, dir, return -errno) {
+                _cleanup_free_ char *p = NULL, *dest = NULL;
+                bool found_path = false, found_dest, b = false;
+                int q;
 
-                        nfd = openat(fd, de->d_name, O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-                        if (nfd < 0) {
-                                if (errno == ENOENT)
-                                        continue;
+                dirent_ensure_type(dir, de);
 
-                                if (r == 0)
-                                        r = -errno;
-                                continue;
-                        }
+                if (de->d_type != DT_LNK)
+                        continue;
 
-                        p = path_make_absolute(de->d_name, path);
-                        if (!p) {
-                                safe_close(nfd);
-                                return -ENOMEM;
-                        }
+                /* Acquire symlink name */
+                p = path_make_absolute(de->d_name, dir_path);
+                if (!p)
+                        return -ENOMEM;
 
-                        /* This will close nfd, regardless whether it succeeds or not */
-                        q = find_symlinks_fd(root_dir, i, match_aliases, ignore_same_name, nfd,
-                                             p, config_path, same_name_link);
-                        if (q > 0)
-                                return 1;
+                /* Acquire symlink destination */
+                q = readlink_malloc(p, &dest);
+                if (q == -ENOENT)
+                        continue;
+                if (q < 0) {
                         if (r == 0)
                                 r = q;
+                        continue;
+                }
 
-                } else if (de->d_type == DT_LNK) {
-                        _cleanup_free_ char *p = NULL, *dest = NULL;
-                        bool found_path = false, found_dest, b = false;
-                        int q;
+                /* Make absolute */
+                if (!path_is_absolute(dest)) {
+                        char *x;
 
-                        /* Acquire symlink name */
-                        p = path_make_absolute(de->d_name, path);
-                        if (!p)
+                        x = path_join(root_dir, dest);
```

Comment:
This looks iffy. `dest` is relative to `dir`, not `root_dir`.

Reply:
Again, it was like this previously. Hmm, it seem wrong indeed.

Reply:
Btw, we need dir_path here, so we need to pass it in. Hence `openat()+fdopendir()` in `find_symlinks()` wouldn't bring any savings.